File: heuristics.py
"""Path-based commit classifier. Fires only when every staged file fits one
category, so mixed commits fall through to the ML model."""
import re
from typing import List, Optional, Tuple

DOC_PATTERNS = [
    re.compile(r".*\.(md|mdx|rst|adoc)$", re.I),
    re.compile(
        r"(^|/)(README|CHANGELOG|CONTRIBUTING|AUTHORS|LICENSE|NOTICE|"
        r"CODE_OF_CONDUCT|SECURITY|MAINTAINERS)(\.[^/]*)?$",
        re.I,
    ),
    re.compile(r"(^|/)docs?/", re.I),
    re.compile(r"(^|/)documentation/", re.I),
]

TEST_PATTERNS = [
    re.compile(r"(^|/)tests?/"),
    re.compile(r"(^|/)__tests__/"),
    re.compile(r"(^|/)spec/"),
    re.compile(r"(^|/)e2e/"),
    re.compile(r"(^|/)test_[^/]+\.py$"),
    re.compile(r".*_test\.(py|go|rb)$"),
    re.compile(r".*\.(test|spec)\.(ts|tsx|js|jsx|mjs|cjs)$"),
    re.compile(r".*Test\.java$"),
    re.compile(r".*Tests?\.cs$"),
    re.compile(r".*_spec\.rb$"),
]

CI_PATTERNS = [
    re.compile(r"^\.github/(workflows|actions)/"),
    re.compile(r"^\.gitlab-ci\.ya?ml$"),
    re.compile(r"^\.circleci/"),
    re.compile(r"^azure-pipelines.*\.ya?ml$"),
    re.compile(r"^Jenkinsfile$"),
    re.compile(r"^\.travis\.ya?ml$"),
    re.compile(r"^\.drone\.ya?ml$"),
    re.compile(r"^\.buildkite/"),
    re.compile(r"^buildkite\.ya?ml$"),
    re.compile(r"^appveyor\.ya?ml$"),
    re.compile(r"^codecov\.ya?ml$"),
    re.compile(r"^\.pre-commit-config\.ya?ml$"),
]


def _all_match(files: List[str], patterns: List[re.Pattern]) -> bool:
    if not files:
        return False
    return all(any(p.search(f) for p in patterns) for f in files)


def heuristic_classify(files: List[str]) -> Optional[Tuple[str, str]]:
    """Return (label, reason) when all files match one high-precision rule."""
    if not files:
        return None
    if _all_match(files, CI_PATTERNS):
        return ("ci", f"all {len(files)} file(s) under CI config paths")
    if _all_match(files, DOC_PATTERNS):
        return ("docs", f"all {len(files)} file(s) are documentation")
    if _all_match(files, TEST_PATTERNS):
        return ("test", f"all {len(files)} file(s) are test files")
    return None

File: gca-rs/src/heuristics.rs
// Path-based pre-classifier. Fires only when every staged file matches the
// same category, so mixed commits fall through to the ML model.
use once_cell::sync::Lazy;
use regex::Regex;

pub struct Hit {
    pub label: &'static str,
    pub reason: String,
}

static DOC: Lazy<Vec<Regex>> = Lazy::new(|| {
    [
        r"(?i).*\.(md|mdx|rst|adoc)$",
        r"(?i)(^|/)(README|CHANGELOG|CONTRIBUTING|AUTHORS|LICENSE|NOTICE|CODE_OF_CONDUCT|SECURITY|MAINTAINERS)(\.[^/]*)?$",
        r"(?i)(^|/)docs?/",
        r"(?i)(^|/)documentation/",
    ]
    .iter()
    .map(|p| Regex::new(p).unwrap())
    .collect()
});

static TEST: Lazy<Vec<Regex>> = Lazy::new(|| {
    [
        r"(^|/)tests?/",
        r"(^|/)__tests__/",
        r"(^|/)spec/",
        r"(^|/)e2e/",
        r"(^|/)test_[^/]+\.py$",
        r".*_test\.(py|go|rb)$",
        r".*\.(test|spec)\.(ts|tsx|js|jsx|mjs|cjs)$",
        r".*Test\.java$",
        r".*Tests?\.cs$",
        r".*_spec\.rb$",
    ]
    .iter()
    .map(|p| Regex::new(p).unwrap())
    .collect()
});

static CI: Lazy<Vec<Regex>> = Lazy::new(|| {
    [
        r"^\.github/(workflows|actions)/",
        r"^\.gitlab-ci\.ya?ml$",
        r"^\.circleci/",
        r"^azure-pipelines.*\.ya?ml$",
        r"^Jenkinsfile$",
        r"^\.travis\.ya?ml$",
        r"^\.drone\.ya?ml$",
        r"^\.buildkite/",
        r"^buildkite\.ya?ml$",
        r"^appveyor\.ya?ml$",
        r"^codecov\.ya?ml$",
        r"^\.pre-commit-config\.ya?ml$",
    ]
    .iter()
    .map(|p| Regex::new(p).unwrap())
    .collect()
});

fn all_match(files: &[String], patterns: &[Regex]) -> bool {
    !files.is_empty() && files.iter().all(|f| patterns.iter().any(|p| p.is_match(f)))
}

pub fn classify(files: &[String]) -> Option<Hit> {
    if files.is_empty() {
        return None;
    }
    if all_match(files, &CI) {
        return Some(Hit { label: "ci", reason: format!("all {} file(s) under CI config paths", files.len()) });
    }
    if all_match(files, &DOC) {
        return Some(Hit { label: "docs", reason: format!("all {} file(s) are documentation", files.len()) });
    }
    if all_match(files, &TEST) {
        return Some(Hit { label: "test", reason: format!("all {} file(s) are test files", files.len()) });
    }
    None
}
